File: pipeline/course_detection.py
"""Rectified map image -> course overprint (controls, start, finish, legs).

The printed course color is NOT a fixed hue across maps (checked against all
9 test photos): red, dark maroon, and pink/magenta all show up depending on
the event. This module casts a broad warm-hue net for "course ink" and then
classifies connected components by *shape* (circle/triangle/thin-line), not
by exact color, which is what actually stays constant across maps -- see
plan's Context section for the full finding.
"""

from __future__ import annotations

import re
import shutil
from dataclasses import dataclass, field
from pathlib import Path

import cv2
import numpy as np
from shapely.geometry import LineString

from pipeline.config import (
    HOUGH_PARAM2,
    HOUGH_PARAM2_DEFAULT,
    RING_COVERAGE_MIN_FRACTION,
    START_TRIANGLE_AREA_RATIO_RANGE,
    START_TRIANGLE_MIN_EQUILATERAL_SCORE,
)

try:
    import pytesseract
except ImportError:  # pragma: no cover
    pytesseract = None

if pytesseract is not None and shutil.which("tesseract") is None:
    # PATH may not include the install dir in every process/session even
    # though the binary is present (verified locally: winget installs it
    # here but doesn't broadcast the PATH update to already-running shells).
    _default_win_install = Path(r"C:\Program Files\Tesseract-OCR\tesseract.exe")
    if _default_win_install.exists():
        pytesseract.pytesseract.tesseract_cmd = str(_default_win_install)


@dataclass
class Control:
    x: float
    y: float
    radius: float
    code: str | None = None  # OCR'd control code, e.g. "46" from a "7-46" label; None if unread


@dataclass
class CourseResult:
    controls: list[Control] = field(default_factory=list)
    start: tuple[float, float] | None = None
    finish: tuple[float, float] | None = None
    legs: list[LineString] = field(default_factory=list)
    ink_mask: np.ndarray | None = None  # exposed so segmentation can exclude these pixels from terrain


def build_course_ink_mask(img: np.ndarray, valid_mask: np.ndarray) -> np.ndarray:
    """Broad warm red/maroon/magenta mask -- covers every course-overprint
    color seen across the test photos (see module docstring). Deliberately
    wide; shape-based classification downstream is what separates real
    course marks from other warm-colored map content."""
    hsv = cv2.cvtColor(img, cv2.COLOR_BGR2HSV)
    h, s, v = cv2.split(hsv)
    red_low = (h <= 15) | (h >= 150)
    warm_enough = (s > 70) & (v > 40) & (v < 235)
    mask = red_low & warm_enough & valid_mask
    mask_u8 = (mask.astype(np.uint8)) * 255
    mask_u8 = cv2.morphologyEx(mask_u8, cv2.MORPH_CLOSE, np.ones((3, 3), np.uint8))
    return mask_u8 > 0


def detect_controls(mask: np.ndarray, img_shape: tuple[int, int], param2: int) -> list[Control]:
    """Control circles: hollow rings of a fairly consistent diameter across
    the whole map (the diameter is a fixed mm value at the map's print
    scale). HoughCircles on the ink mask directly, rather than
    contour/circularity analysis, since a thin ring's own outer contour is a
    circle already and Hough is the more standard/robust tool for that.

    No single param2 generalizes across photos (see config.HOUGH_PARAM2) --
    checked directly against map0.jpg and map2.jpg's manually-counted
    baselines, and a threshold that's about right for one is off by 2-4x on
    the other, because one photo's course-ink mask is simply noisier than
    the other's. A modal-radius consistency filter (real controls share one
    print diameter) was tried as a way to avoid needing a per-file value,
    but measured worse in practice -- Hough's radius estimates on these
    photos are noisy enough (std ~13px on map0.jpg, no sharp peak) that the
    mode just picks a different noisy cluster, not a clean signal. Simple
    near-duplicate suppression is kept below since that part did help.
    Caller is expected to pass config.HOUGH_PARAM2.get(filename,
    HOUGH_PARAM2_DEFAULT).
    """
    h, w = img_shape[:2]
    diag = np.hypot(h, w)
    mask_u8 = (mask.astype(np.uint8)) * 255
    # The sheet's own printed border is a long, strong, near-circular-at-corners
    # edge that would otherwise compete with real control circles.
    mx, my = int(w * 0.015), int(h * 0.015)
    mask_u8[:my, :] = 0
    mask_u8[-my:, :] = 0
    mask_u8[:, :mx] = 0
    mask_u8[:, -mx:] = 0
    blurred = cv2.GaussianBlur(mask_u8, (5, 5), 0)
    circles = cv2.HoughCircles(
        blurred, cv2.HOUGH_GRADIENT, dp=1.2, minDist=diag * 0.02,
        param1=60, param2=param2, minRadius=int(diag * 0.006), maxRadius=int(diag * 0.02),
    )
    if circles is None:
        return []

    out = []
    for x, y, r in circles[0]:
        # cv2.HoughCircles returns candidates ordered by accumulator
        # strength (best first); skip anything too close to an already-kept
        # circle instead of double-detecting the same ring.
        if any(np.hypot(x - c.x, y - c.y) < r for c in out):
            continue
        if _ring_ink_coverage(mask, x, y, r) < RING_COVERAGE_MIN_FRACTION:
            continue
        out.append(Control(x=float(x), y=float(y), radius=float(r)))
    return out


def _ring_ink_coverage(mask: np.ndarray, x: float, y: float, r: float, band: int = 3, samples: int = 144) -> float:
    """Fraction of a candidate circle's own circumference that actually has
    ink under it, sampled every 2.5 degrees and tolerant of the Hough
    radius estimate being off by a few px (band). A real printed control
    ring is one continuous stroke, so a correct detection's circle traces
    ink almost the whole way around; a HoughCircles false positive fit
    through scattered non-ring ink -- clutter, or (checked directly on
    map0.jpg) the round "0" glyphs in a nearby control's own printed code
    label -- only grazes ink along a fraction of its circumference. Checked
    directly against every accepted/rejected circle on all four
    IN_SCOPE_FILES: real rings scored >=0.82, every confirmed false
    positive (including two that a same-file median-radius comparison
    couldn't distinguish, since their radii sat inside the real cluster's
    own range) scored <=0.72 -- see config.RING_COVERAGE_MIN_FRACTION."""
    h, w = mask.shape
    hits = 0
    for i in range(samples):
        theta = 2 * np.pi * i / samples
        for dr in range(-band, band + 1):
            px = int(x + (r + dr) * np.cos(theta))
            py = int(y + (r + dr) * np.sin(theta))
            if 0 <= px < w and 0 <= py < h and mask[py, px]:
                hits += 1
                break
    return hits / samples


def _dedupe_finish(controls: list[Control], dist_tol_frac: float = 0.5) -> tuple[list[Control], tuple[float, float] | None]:
    """Finish is drawn as two concentric circles. Hough will report both as
    separate detections with close centers and different radii -- merge the
    closest such pair into a single finish point and drop both from the
    control list."""
    for i, a in enumerate(controls):
        for j, b in enumerate(controls):
            if i >= j:
                continue
            dist = np.hypot(a.x - b.x, a.y - b.y)
            if dist < max(a.radius, b.radius) * dist_tol_frac and abs(a.radius - b.radius) > 1.5:
                finish = ((a.x + b.x) / 2.0, (a.y + b.y) / 2.0)
                remaining = [c for k, c in enumerate(controls) if k not in (i, j)]
                return remaining, finish
    return controls, None


def detect_start_triangle(mask: np.ndarray, exclude_disks: list[Control]) -> tuple[float, float] | None:
    """Start is a triangle (not a circle) in the same ink. Searches connected
    components of the ink mask, excluding pixels already claimed by a
    detected control/finish circle, for one that approximates a 3-vertex
    polygon with roughly equal sides *and* a size consistent with this
    file's own detected control-circle scale (config.START_TRIANGLE_*).

    This reliably returns None on every in-scope photo as of this writing --
    that's the honest, checked-directly result, not a bug to chase. Five
    separate approaches were tried to make this actually find the real
    triangle, and all five failed for concrete, verified reasons:

    1. The original version of this function (no size gate, equilateral
       threshold 0.6) does return *a* 3-vertex contour, but cropping the
       photo at its reported coordinates shows plain terrain, no triangle
       ink -- it's noise. The "triangles" it accepts are 30-220px^2 (a real
       triangle at these files' print scale should be ~2000-4000px^2, per
       START_TRIANGLE_AREA_RATIO_RANGE's comment) with visibly skewed sides
       (e.g. 44.7/25.7/23.6px) that the old 0.6 threshold didn't reject.
    2. Recalibrating the warm-hue mask tighter, using HSV sampled directly
       from this file's own confirmed control-circle ink as the reference:
       no improvement. Sampled real ring ink and known false positives
       (a printed road, a contour-line tangle, a sponsor logo) side by side
       -- their HSV percentiles are statistically indistinguishable (e.g.
       H median 7 vs 8-17, S median 116-140 vs 68-127 across all three,
       heavily overlapping). These elements are printed in genuinely similar
       red/brown ink families; no threshold separates them by color.
    3. Local illumination flattening (dividing by a heavily-blurred copy per
       channel, the document-scanner "remove shading/vignette" trick) before
       re-sampling: same result, no separation opens up. This rules out
       "it's just a bad white balance" -- the confusion is in the printed
       ink colors themselves, not photo lighting quality.
    4. Comparing local stroke thickness (2x distance-transform) of confirmed
       control-ring ink against the rest of the mask: medians match (4px
       both) -- course lines and contour/road lines print at the same
       weight on these maps, no thickness signal either.
    5. Rotation-invariant template matching (a synthetic hollow triangle at
       the expected size, tested at 15-degree steps): found a strong,
       120-degree-symmetric peak -- genuinely promising until the matched
       location was cropped. It was a straight course leg crossing a
       V-shaped kink in an out-of-bounds/area-boundary line, not a triangle;
       every top candidate was the same kind of incidental line crossing.

    Manually scanning the full rectified photo (including a saturation-
    boosted crop and every config.LEGEND_EXCLUDE_BOXES region, in case the
    triangle was hidden under a hand-excluded logo box) didn't turn up an
    obvious separate triangle mark either. Conclusion: on these photos, at
    this resolution, the real start triangle is not reliably distinguishable
    from ordinary map linework by color, shape, size, thickness, or template
    matching -- reaching it would need a fundamentally different approach
    (e.g. a trained symbol classifier), which is Phase-3 ML-segmentation
    territory, not a Phase-0 CV heuristic. Don't re-attempt any of the five
    approaches above without new evidence for why they'd work differently
    next time (see plan's guidance on this pattern, also applied to the
    abandoned OCR-declutter and radius-consistency-filter attempts).
    """
    work = mask.copy()
    for c in exclude_disks:
        cv2.circle(work, (int(c.x), int(c.y)), int(c.radius * 1.4), 0, thickness=-1)
    mask_u8 = (work.astype(np.uint8)) * 255
    contours, _ = cv2.findContours(mask_u8, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)

    if not exclude_disks:
        return None  # no controls detected to calibrate the expected triangle size against

    median_circle_area = float(np.median([np.pi * c.radius ** 2 for c in exclude_disks]))
    min_area = START_TRIANGLE_AREA_RATIO_RANGE[0] * median_circle_area
    max_area = START_TRIANGLE_AREA_RATIO_RANGE[1] * median_circle_area

    best = None
    best_score = -1.0
    for c in contours:
        area = cv2.contourArea(c)
        if area < min_area or area > max_area:
            continue
        peri = cv2.arcLength(c, True)
        approx = cv2.approxPolyDP(c, 0.04 * peri, True)
        if len(approx) != 3:
            continue
        pts = approx.reshape(3, 2).astype(float)
        sides = [np.hypot(*(pts[i] - pts[(i + 1) % 3])) for i in range(3)]
        equilateral = 1.0 - (np.std(sides) / max(np.mean(sides), 1e-6))
        if equilateral > best_score:
            best_score = equilateral
            best = pts.mean(axis=0)
    if best is None or best_score < START_TRIANGLE_MIN_EQUILATERAL_SCORE:
        return None
    return float(best[0]), float(best[1])


def detect_legs(mask: np.ndarray, exclude_disks: list[Control], start: tuple[float, float] | None) -> list[LineString]:
    """Connecting lines between controls: what's left of the ink mask once
    circles/triangle are carved out is thin line segments -- detect them the
    same way as segmentation.py's path lines (short HoughLinesP segments,
    draft quality, not merged into single polylines per leg)."""
    work = mask.copy()
    for c in exclude_disks:
        cv2.circle(work, (int(c.x), int(c.y)), int(c.radius * 1.4), 0, thickness=-1)
    if start is not None:
        cv2.circle(work, (int(start[0]), int(start[1])), int(mask.shape[0] * 0.02), 0, thickness=-1)
    # Same border-frame exclusion as detect_controls, and for the same
    # reason: checked directly (course_detection false positives, see plan)
    # that the sheet's own printed border ends up in the broad course-ink
    # mask on all four in-scope photos and Hough was tracing long "leg"
    # segments along it, running the full width/height of the sheet.
    h, w = mask.shape[:2]
    my, mx = int(h * 0.015), int(w * 0.015)
    work[:my, :] = False
    work[-my:, :] = False
    work[:, :mx] = False
    work[:, -mx:] = False
    mask_u8 = (work.astype(np.uint8)) * 255
    edges = cv2.Canny(mask_u8, 40, 120)
    # minLineLength scaled to image size and set well above a control-code
    # digit's stroke length (short text like "7-46" next to each circle
    # otherwise floods this with thousands of tiny spurious segments --
    # verified empirically on map0.jpg: minLineLength=12 gives ~5600
    # segments, mostly digit strokes, vs ~90 at this scaled threshold).
    min_len = mask.shape[0] * 0.04
    lines = cv2.HoughLinesP(edges, 1, np.pi / 180, threshold=20, minLineLength=min_len, maxLineGap=15)
    if lines is None:
        return []
    out = []
    for seg in lines:
        x1, y1, x2, y2 = np.asarray(seg).reshape(-1).astype(float)
        if np.hypot(x2 - x1, y2 - y1) < min_len:
            continue
        out.append(LineString([(x1, y1), (x2, y2)]))
    return out


_CODE_PATTERN = re.compile(r"\d{1,2}-\d{1,3}")


def ocr_control_code(img: np.ndarray, control: Control, ink_mask: np.ndarray) -> str | None:
    """OCR the control-code text printed next to a control circle (e.g. the
    "46" in a "7-46" label). Best-effort: requires the Tesseract-OCR binary
    installed separately (pytesseract only binds to it) -- returns None
    rather than raising if it's unavailable or nothing legible is found,
    since this must never fabricate a code that wasn't actually read.

    OCRs the course-ink *mask* within a generous window around the circle,
    not the raw photo -- course setters place the code label in whichever
    of the 4 quadrants around the circle has free space (not a fixed
    offset), and OCRing the raw grayscale crop picks up contour lines and
    vegetation-boundary ink as false text. Restricting to ink-mask pixels
    (with the circle itself blanked out as a filled disk, since Hough's
    radius estimate is noisy enough that a thin ring mask can miss part of
    the real printed circle) cuts that noise down enough for --psm 11
    (sparse text) to isolate the digit-dash label. Checked directly against
    map0.jpg's golden-path photo: this raised the legible-code yield from
    1/17 controls (old fixed-offset raw-grayscale crop) to ~9/17 -- real
    photos are noisy enough that recovering roughly half is the realistic
    best-effort bar here, not full coverage.
    """
    if pytesseract is None:
        return None
    r = control.radius
    h, w = img.shape[:2]
    search = int(r * 6)
    x0, x1 = max(0, int(control.x - search)), min(w, int(control.x + search))
    y0, y1 = max(0, int(control.y - search)), min(h, int(control.y + search))
    roi_mask = ink_mask[y0:y1, x0:x1].copy()
    if roi_mask.size == 0:
        return None
    cv2.circle(roi_mask, (int(control.x - x0), int(control.y - y0)), int(r * 1.6), False, thickness=-1)
    binary = np.where(roi_mask, 0, 255).astype(np.uint8)
    binary = cv2.resize(binary, None, fx=3, fy=3, interpolation=cv2.INTER_NEAREST)
    binary = cv2.medianBlur(binary, 3)
    try:
        data = pytesseract.image_to_data(
            binary, config="--psm 11 -c tessedit_char_whitelist=0123456789-",
            output_type=pytesseract.Output.DICT,
        )
    except Exception:
        return None
    best_code, best_conf = None, -1.0
    for text, conf in zip(data["text"], data["conf"]):
        match = _CODE_PATTERN.search(text.strip())
        if match and float(conf) > best_conf:
            best_code, best_conf = match.group(0), float(conf)
    return best_code


def detect_course(
    img: np.ndarray, valid_mask: np.ndarray, run_ocr: bool = True, source_filename: str | None = None,
) -> CourseResult:
    param2 = HOUGH_PARAM2.get(source_filename, HOUGH_PARAM2_DEFAULT)
    mask = build_course_ink_mask(img, valid_mask)
    controls = detect_controls(mask, img.shape, param2)
    controls, finish = _dedupe_finish(controls)
    start = detect_start_triangle(mask, controls)
    legs = detect_legs(mask, controls, start)

    if run_ocr:
        for c in controls:
            c.code = ocr_control_code(img, c, mask)

    return CourseResult(controls=controls, start=start, finish=finish, legs=legs, ink_mask=mask)

File: app/rust/src/api/course_detection.rs
//! Rectified map image -> course overprint (controls, start, finish, legs).
//! Conceptual Rust port of the Phase 0 Python prototype's
//! `course_detection.py` -- see its module docstring for why course color is
//! deliberately NOT treated as a fixed hue (red/maroon/magenta all show up
//! depending on the event): this casts a broad warm-hue net for "course ink"
//! and classifies connected components by *shape* (circle/triangle/
//! thin-line), which is what actually stays constant across maps.
//!
//! OCR of printed control codes (`ocr_control_code` in the Python version)
//! is NOT ported here -- it depends on a separately-installed Tesseract
//! binary, which is a distinct infrastructure decision for Android/Flutter
//! (bundling or requiring a native OCR engine) rather than something to pull
//! in incidentally while porting shape detection. `Control::code` is always
//! `None` for now.

use opencv::core::{self, Mat, Point, Scalar, Vector};
use opencv::prelude::*;
use opencv::imgproc;

use super::geometry::{Pt, Segment};
use super::mask_utils::{mask_to_mat, mat_to_mask};
use super::segmentation::hough_segments;

#[derive(Debug, Clone)]
pub struct Control {
    pub x: f32,
    pub y: f32,
    pub radius: f32,
    /// OCR'd control code, e.g. "46" from a "7-46" label. Always `None` for
    /// now -- see module docs.
    pub code: Option<String>,
}

#[derive(Debug, Clone, Default)]
pub struct CourseResult {
    pub controls: Vec<Control>,
    pub start: Option<Pt>,
    pub finish: Option<Pt>,
    pub legs: Vec<Segment>,
}

/// Broad warm red/maroon/magenta mask -- covers every course-overprint color
/// seen across the Python prototype's test photos. Deliberately wide;
/// shape-based classification downstream is what separates real course
/// marks from other warm-colored map content.
pub(crate) fn build_course_ink_mask(img: &Mat, valid_mask: &[u8]) -> opencv::Result<Vec<u8>> {
    let (width, height) = (img.cols(), img.rows());
    let mut hsv = Mat::default();
    imgproc::cvt_color_def(img, &mut hsv, imgproc::COLOR_BGR2HSV)?;
    let hsv_bytes = hsv.data_bytes()?;

    let mut mask = vec![0u8; (width * height) as usize];
    for i in 0..mask.len() {
        let (h, s, v) = (hsv_bytes[i * 3], hsv_bytes[i * 3 + 1], hsv_bytes[i * 3 + 2]);
        let red_low = h <= 15 || h >= 150;
        let warm_enough = s > 70 && v > 40 && v < 235;
        if red_low && warm_enough && valid_mask[i] != 0 {
            mask[i] = 255;
        }
    }

    let mat = mask_to_mat(&mask, width, height)?;
    let kernel = Mat::new_rows_cols_with_default(3, 3, core::CV_8U, Scalar::all(1.0))?;
    let mut closed = Mat::default();
    imgproc::morphology_ex_def(&mat, &mut closed, imgproc::MORPH_CLOSE, &kernel)?;
    mat_to_mask(&closed)
}

/// Fraction of a candidate circle's own circumference that actually has ink
/// under it, sampled every 2.5 degrees and tolerant of the Hough radius
/// estimate being off by a few px (`band`). Ported as-is from the Python
/// prototype's `_ring_ink_coverage` -- see `RING_COVERAGE_MIN_FRACTION`'s
/// caller-side comment for why this, not a radius-consistency filter, is
/// what actually separates real control rings from Hough false positives.
fn ring_ink_coverage(mask: &[u8], width: i32, height: i32, x: f32, y: f32, r: f32) -> f32 {
    const BAND: i32 = 3;
    const SAMPLES: i32 = 144;
    let mut hits = 0;
    for i in 0..SAMPLES {
        let theta = 2.0 * std::f32::consts::PI * i as f32 / SAMPLES as f32;
        for dr in -BAND..=BAND {
            let px = (x + (r + dr as f32) * theta.cos()) as i32;
            let py = (y + (r + dr as f32) * theta.sin()) as i32;
            if px >= 0 && px < width && py >= 0 && py < height && mask[(py * width + px) as usize] != 0 {
                hits += 1;
                break;
            }
        }
    }
    hits as f32 / SAMPLES as f32
}

/// Control circles: hollow rings of a fairly consistent diameter across the
/// whole map. `param2` is the caller-supplied HoughCircles accumulator
/// threshold -- deliberately not a hardcoded per-file table here (see the
/// Python prototype's `config.HOUGH_PARAM2` and `PHASE0_HANDOFF.md`'s
/// guidance that it doesn't generalize to an unseen photo and belongs behind
/// a manual-correction/sensitivity-tuning UI control instead).
pub(crate) fn detect_controls(
    mask: &[u8],
    width: i32,
    height: i32,
    param2: f64,
    ring_coverage_min_fraction: f32,
) -> opencv::Result<Vec<Control>> {
    let diag = ((width * width + height * height) as f64).sqrt();
    let mut masked = mask.to_vec();
    // The sheet's own printed border is a long, strong, near-circular-at-
    // corners edge that would otherwise compete with real control circles.
    let (mx, my) = ((width as f64 * 0.015) as i32, (height as f64 * 0.015) as i32);
    for y in 0..height {
        for x in 0..width {
            if y < my || y >= height - my || x < mx || x >= width - mx {
                masked[(y * width + x) as usize] = 0;
            }
        }
    }
    let mask_mat = mask_to_mat(&masked, width, height)?;
    let mut blurred = Mat::default();
    imgproc::gaussian_blur_def(&mask_mat, &mut blurred, core::Size::new(5, 5), 0.0)?;

    let mut circles = Vector::<core::Vec3f>::new();
    imgproc::hough_circles(
        &blurred,
        &mut circles,
        imgproc::HOUGH_GRADIENT,
        1.2,
        diag * 0.02,
        60.0,
        param2,
        (diag * 0.006) as i32,
        (diag * 0.02) as i32,
    )?;

    let mut out: Vec<Control> = Vec::new();
    for c in circles.iter() {
        let (x, y, r) = (c[0], c[1], c[2]);
        // HoughCircles returns candidates ordered by accumulator strength
        // (best first); skip anything too close to an already-kept circle
        // instead of double-detecting the same ring.
        if out.iter().any(|k| ((x - k.x).powi(2) + (y - k.y).powi(2)).sqrt() < r) {
            continue;
        }
        if ring_ink_coverage(mask, width, height, x, y, r) < ring_coverage_min_fraction {
            continue;
        }
        out.push(Control { x, y, radius: r, code: None });
    }
    Ok(out)
}

/// Finish is drawn as two concentric circles. Hough reports both as
/// separate detections with close centers and different radii -- merge the
/// closest such pair into a single finish point and drop both from the
/// control list.
fn dedupe_finish(controls: Vec<Control>) -> (Vec<Control>, Option<Pt>) {
    for i in 0..controls.len() {
        for j in (i + 1)..controls.len() {
            let (a, b) = (&controls[i], &controls[j]);
            let dist = ((a.x - b.x).powi(2) + (a.y - b.y).powi(2)).sqrt();
            if dist < a.radius.max(b.radius) * 0.5 && (a.radius - b.radius).abs() > 1.5 {
                let finish = Pt { x: (a.x + b.x) / 2.0, y: (a.y + b.y) / 2.0 };
                let remaining =
                    controls.iter().enumerate().filter(|&(k, _)| k != i && k != j).map(|(_, c)| c.clone()).collect();
                return (remaining, Some(finish));
            }
        }
    }
    (controls, None)
}

fn blank_disks(mask: &[u8], width: i32, height: i32, disks: &[(f32, f32, f32)]) -> opencv::Result<Mat> {
    let mut mat = mask_to_mat(mask, width, height)?;
    for &(x, y, r) in disks {
        imgproc::circle(&mut mat, Point::new(x as i32, y as i32), r as i32, Scalar::all(0.0), -1, imgproc::LINE_8, 0)?;
    }
    Ok(mat)
}

/// Start is a triangle (not a circle) in the same ink. This reliably returns
/// `None` on the Python prototype's in-scope photos -- see
/// `course_detection.detect_start_triangle`'s docstring in
/// `python_prototype/pipeline/course_detection.py` for the full account of
/// five separate approaches that were tried and failed for concrete,
/// verified reasons (an unguarded shape search finds noise, not a real
/// triangle; tighter HSV recalibration, illumination flattening, and
/// stroke-thickness comparison show no separating signal; rotation-invariant
/// template matching found a false peak on a line-crossing, not a triangle).
/// Reliably finding it would need a trained symbol classifier -- Phase 3 ML
/// territory, not a Phase 0/1 CV heuristic. This port keeps the same
/// size/equilateral-ness gate so the behavior (usually `None`) matches, not
/// because the gate is expected to start working better in Rust.
pub(crate) fn detect_start_triangle(
    mask: &[u8],
    width: i32,
    height: i32,
    controls: &[Control],
    area_ratio_range: (f64, f64),
    min_equilateral_score: f64,
) -> opencv::Result<Option<Pt>> {
    if controls.is_empty() {
        return Ok(None);
    }
    let disks: Vec<(f32, f32, f32)> = controls.iter().map(|c| (c.x, c.y, c.radius * 1.4)).collect();
    let work = blank_disks(mask, width, height, &disks)?;

    let mut contours = Vector::<Vector<Point>>::new();
    imgproc::find_contours_def(&work, &mut contours, imgproc::RETR_EXTERNAL, imgproc::CHAIN_APPROX_SIMPLE)?;

    let mut circle_areas: Vec<f64> = controls.iter().map(|c| std::f64::consts::PI * (c.radius as f64).powi(2)).collect();
    circle_areas.sort_by(|a, b| a.partial_cmp(b).unwrap());
    let median_circle_area = circle_areas[circle_areas.len() / 2];
    let min_area = area_ratio_range.0 * median_circle_area;
    let max_area = area_ratio_range.1 * median_circle_area;

    let mut best: Option<Pt> = None;
    let mut best_score = -1.0f64;
    for c in contours.iter() {
        let area = imgproc::contour_area_def(&c)?;
        if area < min_area || area > max_area {
            continue;
        }
        let peri = imgproc::arc_length(&c, true)?;
        let mut approx = Vector::<Point>::new();
        imgproc::approx_poly_dp(&c, &mut approx, 0.04 * peri, true)?;
        if approx.len() != 3 {
            continue;
        }
        let pts: Vec<(f64, f64)> = approx.iter().map(|p| (p.x as f64, p.y as f64)).collect();
        let sides: Vec<f64> = (0..3)
            .map(|i| {
                let (ax, ay) = pts[i];
                let (bx, by) = pts[(i + 1) % 3];
                ((ax - bx).powi(2) + (ay - by).powi(2)).sqrt()
            })
            .collect();
        let mean = sides.iter().sum::<f64>() / 3.0;
        let variance = sides.iter().map(|s| (s - mean).powi(2)).sum::<f64>() / 3.0;
        let std = variance.sqrt();
        let equilateral = 1.0 - (std / mean.max(1e-6));
        if equilateral > best_score {
            best_score = equilateral;
            let cx = pts.iter().map(|p| p.0).sum::<f64>() / 3.0;
            let cy = pts.iter().map(|p| p.1).sum::<f64>() / 3.0;
            best = Some(Pt { x: cx as f32, y: cy as f32 });
        }
    }
    if best_score < min_equilateral_score {
        return Ok(None);
    }
    Ok(best)
}

/// Connecting lines between controls: what's left of the ink mask once
/// circles/triangle are carved out is thin line segments -- detected the
/// same way as `segmentation`'s path lines.
pub(crate) fn detect_legs(
    mask: &[u8],
    width: i32,
    height: i32,
    controls: &[Control],
    start: Option<Pt>,
) -> opencv::Result<Vec<Segment>> {
    let mut disks: Vec<(f32, f32, f32)> = controls.iter().map(|c| (c.x, c.y, c.radius * 1.4)).collect();
    if let Some(s) = start {
        disks.push((s.x, s.y, height as f32 * 0.02));
    }
    let mut work = blank_disks(mask, width, height, &disks)?;

    // Same border-frame exclusion as detect_controls, and for the same
    // reason (the sheet's own printed border otherwise gets traced as a
    // spurious "leg" running the full width/height of the sheet).
    let (mx, my) = ((width as f64 * 0.015) as i32, (height as f64 * 0.015) as i32);
    imgproc::rectangle(
        &mut work,
        core::Rect::new(0, 0, width, my),
        Scalar::all(0.0),
        -1,
        imgproc::LINE_8,
        0,
    )?;
    imgproc::rectangle(
        &mut work,
        core::Rect::new(0, height - my, width, my),
        Scalar::all(0.0),
        -1,
        imgproc::LINE_8,
        0,
    )?;
    imgproc::rectangle(&mut work, core::Rect::new(0, 0, mx, height), Scalar::all(0.0), -1, imgproc::LINE_8, 0)?;
    imgproc::rectangle(
        &mut work,
        core::Rect::new(width - mx, 0, mx, height),
        Scalar::all(0.0),
        -1,
        imgproc::LINE_8,
        0,
    )?;

    let mut edges = Mat::default();
    imgproc::canny_def(&work, &mut edges, 40.0, 120.0)?;
    // minLineLength scaled to image size, well above a control-code digit's
    // stroke length -- see the Python prototype's detect_legs comment on why
    // (short text like "7-46" next to each circle otherwise floods this with
    // thousands of spurious tiny segments).
    let min_len = height as f64 * 0.04;
    hough_segments(&edges, 20, min_len, 15.0)
}

/// Runs control/start/finish/leg detection against an already-computed
/// course-ink mask (`build_course_ink_mask`) -- the caller builds that mask
/// once and reuses it both here and to exclude course ink from
/// `segmentation`'s terrain classification (see the Python prototype's
/// `CourseResult.ink_mask` field, which exists for exactly that reason).
pub(crate) fn detect_course(
    mask: &[u8],
    width: i32,
    height: i32,
    hough_param2: f64,
    ring_coverage_min_fraction: f32,
    start_triangle_area_ratio_range: (f64, f64),
    start_triangle_min_equilateral_score: f64,
) -> opencv::Result<CourseResult> {
    let controls = detect_controls(mask, width, height, hough_param2, ring_coverage_min_fraction)?;
    let (controls, finish) = dedupe_finish(controls);
    let start = detect_start_triangle(
        mask,
        width,
        height,
        &controls,
        start_triangle_area_ratio_range,
        start_triangle_min_equilateral_score,
    )?;
    let legs = detect_legs(mask, width, height, &controls, start)?;

    Ok(CourseResult { controls, start, finish, legs })
}
